### ISSUE TITLE
add support for Lombok private constructor

### ISSUE THREAD

**Issue by Schattensilber** [checkstyle/checkstyle#2579]
http://jnb.ociweb.com/jnb/jnbJan2010.html
https://projectlombok.org/api/lombok/NoArgsConstructor.html
https://projectlombok.org/features/Constructor.html

```
/var/tmp $ cat Test.java 
import lombok.AccessLevel;
import lombok.NoArgsConstructor;

@NoArgsConstructor(access = AccessLevel.PRIVATE) //<-- checkstyle warning
public final class Foo {
  static void foo() { }
}

/var/tmp $ cat Test.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name = "Checker">
    <module name="TreeWalker">
        <module name="HideUtilityClassConstructor" />
    </module>
</module>

/var/tmp $ java -jar checkstyle-6.13-all.jar -c Test.xml Test.java
Starting audit...
/var/tmp/Test.java:4:1: error: Utility classes should not have a public or default constructor.
Audit done.
Checkstyle ends with 1 errors.
```
##

**Comment by romani** [comment]
Thanks for reporting the issue, most likely we will not support this, but I will think more about this case.

Please use suppressions for such cases  , they are so specific .

**Comment by romani** [comment]
from specification

> @NoArgsConstructor will generate a constructor with no parameters. If this is not possible (because of final fields), a compiler error will result instead, unless 

Public c-tor will be generated, but Check demand it to be private.
Checkstyle works only on text sources, not a compiled binaries - we can not support Lombok.

Suppression/Filters is the only way to avoid this violation - http://checkstyle.sourceforge.net/config_filters.html.

**Comment by broglea** [comment]
I know you closed this issue because you said CheckStyle only works on text sources not a compiled binary. The proper way to use the `@NoArgsConstructor` for a utility class that _should_ pass CheckStyle is with `@NoArgsConstructor(access = AccessLevel.PRIVATE)`. I believe this could definitely be checked by CheckStyle and I for one would definitely appreciate and welcome that change. Since it is very specific to using Lombok, I understand if you don't want to support it but while googling around for how to handle this case I came across this and wanted to give some input as it can be checked from a text source and made private.

**Comment by romani** [comment]
All we can do is too make additional option for HideUtilityClassConstructor to skip validation on classes that have specific annotation. New option `ignoreAnnotationCanonicalNames` , example of processing could be found at  https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/design/VisibilityModifierCheck.java#L53 .

Please be welcome with PR, as we have too much issues and PRs, it is unlikely that we will find time for this issue in next several months.

**Comment by lc-nyovchev** [comment]
would be nice to work for both 

`@NoArgsConstructor(access = AccessLevel.PRIVATE)` and for `@UtilityClass`

**Comment by romani** [comment]
New option will be a list of names, so user could define set of annotations.

**Comment by lc-nyovchev** [comment]
@romani yes, that'd be okay as well, and I think this option should by default include the above mentioned annotations from lombok as they are fairly widely used to my knowledge.

**Comment by mikee805** [comment]
adding suppose for lombok annotations would be great

**Comment by pbludov** [comment]
I'm on it. 
This issue requires some refactoring of class `AnnotationUtil` for code reuse.

**Comment by romani** [comment]
As we have xpath support it is better  do suppression by xpath filter http://checkstyle.sourceforge.net/config_filters.html#SuppressionXpathFilter , looks like Check is already supported by this filter.

To make it more convenient to use by users in single config file, it is better to implement single xpath filter, we already have issue on this

**Comment by pbludov** [comment]
@Schattensilber @broglea @lc-nyovchev @mikee805 
Please confirm that this config is working for you:

TestClass.java
```
import lombok.AccessLevel;
import lombok.NoArgsConstructor;
import lombok.UtilityClass;

@NoArgsConstructor(access = AccessLevel.PRIVATE)
public final class GoogClass { // OK
  static void foo() { }
}

@UtilityClass
public final class AlsoGoogClass { // OK
  static void foo() { }
}

public final class BadClass { // warning: Utility classes should not have a public or default constructor.
  static void foo() { }
}
```

TestConfig.xml
```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name = "Checker">
    <module name="TreeWalker">
        <module name="HideUtilityClassConstructor" />
		<module name="SuppressionXpathFilter">
			<property name="file" value="suppressions.xml"/>
			<property name="optional" value="false"/>
		</module>
    </module>
</module>
```
suppressions.xml
```
<?xml version="1.0"?>

<!DOCTYPE suppressions PUBLIC
"-//Checkstyle//DTD SuppressionXpathFilter Experimental Configuration 1.2//EN"
"https://checkstyle.org/dtds/suppressions_1_2_xpath_experimental.dtd">

<suppressions>
  <suppress-xpath checks="HideUtilityClassConstructor"
    query="//CLASS_DEF[MODIFIERS/ANNOTATION[@text='NoArgsConstructor' or @text='UtilityClass']]"/>
</suppressions>
```

The query
`//CLASS_DEF[MODIFIERS/ANNOTATION[@text='NoArgsConstructor' or @text='UtilityClass']]`
means "Any class with annotation `NoArgsConstructor` or annotation `UtilityClass` should match".

This will not work for the full type name:
```
@lombok.UtilityClass
public final class AnotherClass { // Violation
  static void foo() { }
}
```
For this case the query should be extended with ` or @text='lombok.UtilityClass'`.

**Comment by romani** [comment]
Separate file could be avoided ones https://github.com/checkstyle/checkstyle/issues/5879 become implemented

**Comment by lc-nyovchev** [comment]
@romani it's perfect, exactly what we want, especially without the separate class

**Comment by andreaslorenzbm** [comment]
@romani @pbludov When trying to use the SupressionXpathFilter or the SuppressionXpathSingleFilter I get this error:

An error occurred while generating the checkstyle report. Error configuring the checkstyle: The module TreeWalker can not be initialized - Can you instantiate com.puppycrawl.tools.checkstyle.filters.SuppressionXpathFilter 'class, also is not possible to instantiate it as com.puppycrawl.tools.checkstyle.checks.annotation.com .puppycrawl.tools.checkstyle.filters.SuppressionXpathFilter, com.puppycrawl.tools.checkstyle.[...] Please check if you know the names of the language short names http: //checkstyle.sourceforge .net / config.html# Configure packets. Please also check if the ClassLader to Checker is configured correctly. -> [Help 1]

I have already tried to change the Classname. I updated my checkstyle version. But whatever I try to do, I always get this error.
My xml looks exactly the way @pbludov showed it.
Is there any chance you could help me out?

**Comment by sanyarnd** [comment]
To anyone who's looking for solution:
```xml
<module name="HideUtilityClassConstructor"/>
<module name="SuppressionXpathSingleFilter">
	<property name="checks" value="HideUtilityClassConstructor"/>
	<property name="query" value="//*[MODIFIERS//*[@text = 'UtilityClass' or @text='NoArgsConstructor' or @text = 'SpringBootApplication']]/descendant-or-self::node()"/>
</module>
```

**Comment by romani** [comment]
I am closing this issue for now, there is workaround. If user continue to request it to be done by Check only, we can reopen issue and make special property for this.

Users, please keep sharing xpath for suppression if it requires update.
